### DISCUSSION TITLE
Why does cruft accumulate under `_site`?

### DISCUSSION THREAD

**Discussion by mpf** [quarto-dev/quarto-cli#8551] [Q&A]
### Description

I'm using quarto to build a course [website](https://friedlander.io/ubc-cpsc-406-23-24/). I've found that as I repeatedly cycle through `preview`-`render` while editing, the directory `_site` accumulates what seems to be redundant version of files and directories, eg,
```
_site/site_libs
_site/site_libs 2
_site/site_libs 3
```
These additional files seem to correlate with the site rendering incorrectly (eg, missing style files). My workaround is now to always delete `_site` before publishing:
```
rm -rf _site; quarto publish
```

Is there something, say, in the project's `yml` configuration files that may cause this to happen? I unfortunately can't replicate the behavior with a MWE, so here's the [link to the full repo](https://github.com/mpf/ubc-cpsc-406-23-24.git). (I'm also glad to hear any other advice on how to better setup the project.)

Using quarto v1.4.549

Thanks!

**Comment by mcanouil** [comment]
Could you provide the exact steps to reproduce?
For instance, `quarto publish` is not the complete command.

Did you try to `quarto render` then use `quarto publish --no-render`?

Do you have a simpler example? For instance your are using Julia, but you did not define the Jupyter kernel nor the packages needed.

Why are `_freeze` in your `.gitgnore` if you want `freeze: auto`? Same thing with `cache: true`.

**Reply by mpf** [reply]
> For instance, `quarto publish` is not the complete command.
> Did you try to `quarto render` then use `quarto publish --no-render`?

I use `quarto publish gh-pages`. In particular, because the problem I'm having, the sequence I use to publish is

```
rm -rf _site
quarto render # (then verify pages in _site work)
quarto publish gh-pages --no-render
```

> Do you have a simpler example?

Unfortunately, no.

> For instance your are using Julia, but you did not define the Jupyter kernel nor the packages needed.

Is it required to specify the kernel? I was under the impression that the first code block would determine the kernel. With regard to the required Julia packages, these are maintained in `Project.toml`.

> Why are `_freeze` in your `.gitgnore` if you want freeze: auto? Same thing with `cache: true`.

Makes sense --- I'll remove `_freeze` from `.gitignore`. But I don't think that's the problem, because I only use a single machine to edit and render, and so `_freeze` is always current.

I'm not sure where I'm meant to specify format options. Currently, the project consists of html and revealjs format documents:
- html docs in `/` (eg, `index.qmd`) and `notes/*.qmd`
- revealjs docs in `slides/*.qmd` together with `slides/_metadata.yml'

Is it ok to specify html format options in `_quarto.yml` (because they apply to docs in `/` and `notes/`), and specify revealjs format options in `/slides/_metadata.yml`? Will the html format options in `_quarto.yml` inadvertantly propagate to docs under `slides/'?

**Reply by mcanouil** [reply]
Since `quarto publish gh-pages` cleanup the output directory, something must be wrong in your setup.
GitHub Pages uses either the `gh-pages` or the `docs` directory, why have you a `_site` directory exactly?
Edit: so you are publishing to the branch, but from your local environment.

> Is it required to specify the kernel? I was under the impression that the first code block would determine the kernel. With regard to the required Julia packages, these are maintained in Project.toml.

Well, it's reproducibility. I am using Julia 1.10 for instance, but do you?
Even if Quarto guesses or try to, I strongly believe it should be explicit, especially when sharing with others.

> Makes sense --- I'll remove _freeze from .gitignore. But I don't think that's the problem, because I only use a single machine to edit and render, and so _freeze is always current.

Again, reproducibility. Here, I spent few minutes trying to make your project work but was lacking information. Having freeze/cache could help not having to compute anything.

Read <https://quarto.org/docs/projects/quarto-projects.html#shared-metadata> to understand the metadata level.

**Reply by mcanouil** [reply]
I tried to reproduce but was totally unable to do so.

```sh
quarto create project website demo
cd demo
git init
git remote add ...
quarto publish gh-pages --no-prompt --no-browser
quarto publish gh-pages --no-prompt --no-browser
quarto publish gh-pages --no-prompt --no-browser
```

This is the directory tree and the `gh-pages` contains exactly the content of `_site`.

```sh
.
├── _quarto.yml
├── _site
│   ├── about.html
│   ├── index.html
│   ├── robots.txt
│   ├── search.json
│   ├── site_libs
│   │   ├── bootstrap
│   │   │   ├── bootstrap-icons.css
│   │   │   ├── bootstrap-icons.woff
│   │   │   ├── bootstrap.min.css
│   │   │   └── bootstrap.min.js
│   │   ├── clipboard
│   │   │   └── clipboard.min.js
│   │   ├── quarto-html
│   │   │   ├── anchor.min.js
│   │   │   ├── popper.min.js
│   │   │   ├── quarto-syntax-highlighting.css
│   │   │   ├── quarto.js
│   │   │   ├── tippy.css
│   │   │   └── tippy.umd.min.js
│   │   ├── quarto-nav
│   │   │   ├── headroom.min.js
│   │   │   └── quarto-nav.js
│   │   └── quarto-search
│   │       ├── autocomplete.umd.js
│   │       ├── fuse.min.js
│   │       └── quarto-search.js
│   ├── sitemap.xml
│   └── styles.css
├── about.qmd
├── index.qmd
└── styles.css

8 directories, 26 files
```

**Comment by mpf** [comment]
My hypothesis is that there's something about the particular configuration options that's causing the problem. Your attempt to reproduce doesn't seem to have any of the particular aspects of my project, so I don't know that it's at all informative.

Perhaps one of the developers can comment on what are the conditions that cause quarto to trigger multiple versions of `_site/site_libs`. Then I might try to accordingly change the configurations of the project.

**Reply by mcanouil** [reply]
> Your attempt to reproduce doesn't seem to have any of the particular aspects of my project, so I don't know that it's at all informative.

Unfortunately, your project is way too big for us to do the filtering to find out what might be causing this. Here is nothing obvious (at least to me) in your project.
You know your project, so you can remove piece by piece the components of your project until you have something small and reproducible.
There is no magic trick, the above is the process we are all following to debug something.
Again, unfortunately, the team does not have the time to debug every users' projects, thus why we are asking for small reproducible example.

**Reply by mpf** [reply]
Fair enough. I would still be glad for an answer to this question, which may help guide me in fixing the problem:

> Perhaps one of the developers can comment on what are the conditions that cause quarto to trigger multiple versions of _site/site_libs. Then I might try to accordingly change the configurations of the project.
